gear-programs/vft-manager/app/src/services/token_mapping.rs

```diff
@@ -97,12 +90,71 @@ impl TokenMap {
         self.vara_to_eth
             .clone()
             .into_iter()
-            .map(|(vara_token, eth_token)| {
-                let supply = self
-                    .get_supply_type(&vara_token)
-                    .expect("Should be present due to the TokenMap invariants");
-                (vara_token, eth_token, supply)
-            })
+            .map(|(vara_token, (eth_token, supply))| (vara_token, eth_token, supply))
             .collect()
     }
+
+    fn swap_maps(
+        &mut self,
+        vara_to_eth: &mut HashMap<ActorId, (H160, TokenSupply)>,
+        eth_to_vara: &mut HashMap<H160, ActorId>,
+    ) {
+        mem::swap(&mut self.vara_to_eth, vara_to_eth);
+        mem::swap(&mut self.eth_to_vara, eth_to_vara);
+    }
+
+    #[cfg(feature = "mocks")]
+    pub fn calculate_gas_for_token_map_swap(&mut self) {
+        let mut vara_to_eth = Default::default();
+        let mut eth_to_vara = Default::default();
+
+        self.swap_maps(&mut vara_to_eth, &mut eth_to_vara);
+    }
+
+    pub async fn update_vfts(&mut self, gas_required: u64, vft_map: Vec<(ActorId, ActorId)>) {
+        let mut vara_to_eth = HashMap::with_capacity(self.vara_to_eth.len());
+        let mut eth_to_vara = HashMap::with_capacity(vara_to_eth.len());
+
+        let vft_manager = gstd::exec::program_id();
+        let service = extended_vft_client::Vft::new(GStdRemoting);
+        for (vft, (erc20, supply)) in &self.vara_to_eth {
+            let vft_new = match vft_map
+                .iter()
+                .find_map(|(vft_old, vft_new)| (vft_old == vft).then_some(vft_new))
+            {
+                None => vft,
+
+                Some(vft_new) => match service.balance_of(vft_manager).recv(*vft).await {
+                    Ok(_) => vft,
+
+                    Err(e) => {
+                        if !matches!(
+                            e,
+                            SailsError::GStd(GStdError::ErrorReply(
+                                _,
+                                ErrorReplyReason::InactiveActor
+                            ))
+                        ) {
+                            panic!("Vft failed: {e:?}")
+                        } else {
+                            vft_new
```

Comment:
Why is it like that?

Reply:
It's considered that vft should be in the exited state. If vft responses with a successful result then its address is not updated.

For any error other than `InactiveError` we just stop updating process since the VFT is in some inconsistent state and should not be updated but just removed from the map. The admin should beforehand remove the VFT calling `remove_vara_to_eth_address` method.